File: Code/Java/Repeating_And_Missing_Number.java
/*Repeating And Missing Numbers :-
     Given an array of size n.
     Array elements are in the range from 1 to n. One number from set {1, 2, …n} is missing and one number occurs twice in the array.
     we are going to find these two numbers.

  Algorithm
    This method involves creating a Hashset with the help of Sets. In this, the elements are assigned to their natural index. In this process,
    if an element is assigned twice, then it is the repeating element. And if an element’s assigning is not there, then it is the missing element.
*/

package reapeating_and_mising_number;

import java.io.*;
import java.util.*;

public class Repeating_And_Missing_Number {
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("Enter the Number of Intergers : ");
        //Taking the input of n
        int n = Integer.parseInt(br.readLine());   
        
        int[] a; 
        a = new int[n];
        for(int i = 0; i < n; i++) {   
            a[i] = Integer.parseInt(br.readLine());
        } 
        
        int[] ans = findTwoElement(a, n);                       
        System.out.println("Repeating Number"+"  "+ans[0] + " " +"Missing Number"+"  "+ ans[1]);
        
    }
    
    
    public static int[] findTwoElement(int arr[], int n) {
      
       Set<Integer> numSet = new HashSet<Integer>();    
       int[] a = {0,0};

       for(int i = 0; i < n; i++) {
        //Checking if numset contains that element
           if(!numSet.contains(arr[i]))     
           {   
            //if condition gets false then adding the element to numset
            numSet.add(arr[i]);         
           } else {    
            //assigning value to a[0](repeating number)
             a[0] = arr[i];              
           }
       }
       for(int num = 1; num <= n; num++) {  
            //checking for missing number 
           if(!numSet.contains(num))                     
            {
                a[1] = num;
            }
        }
       return a;
         
    }
}

/*
Test Cases

Input : Enter the Number of Intergers:
4
1
2
2
4
Output : Repeating Number  2 Missing Number  3

Input : Enter the Number of Intergers:
5
1
2
2
4
5
Output : Repeating Number  2 Missing Number  3

TIME COMPLEXITY AND SPACE COMPLEXITY
Time Complexity :- O(n), Where n is the size of array.
Space Complexity :- O(1)
*/

File: Code/C++/repeating_and_missing_number.cpp
/*    Repeating and Missing number

An unsorted array of size n is given where the elements in the array range from 1 to n.
One number from 1 to n is missing and one number occurs twice in the array.
We have to find these two numbers.

Algorithm -

1. We will try to place all the elements on their positions such that a[i] = i+1.
2. Take a pointer at the starting of the array.
3. If the current element is at its position, move the pointer to next position.
4. Else place it at its position by swapping
5. Repeat the above steps till whole array is traversed.

*/

#include<bits/stdc++.h>
using namespace std;

int main()
{
    // n is the size of array
    int n;
    cin >> n;

    // taking array as input
    int a[n];
    for ( int i = 0 ; i < n ; i++ )
        cin >> a[i];

    int i = 0;
    while ( i < n )
    {
        // if element is at its position
        if ( a[i] == i + 1 || a[i] == a[ a[i] - 1 ] )
            i++ ;
        else
            // if not swap
            swap ( a[i], a[ a[i] - 1 ] );
    }

    int r = 0, m = 0;
    for ( int i = 0 ; i < n ; i++ )
    {
        // find the element which is not at its position
        if ( a[i] != i + 1 )
        {
            r = a[i];
            m = i + 1;
            break;
        }
    }

    cout << "Repeating number is " << r << endl;
    cout << "Missing number is " << m;
}

/*
Testcases
Input :
5
4 1 3 2 1
Output : Repeating number is 1
Missing number is 5

Input :
6
1 4 5 2 3 4
Output : Repeating number is 4
Missing number is 6

Time complexity : O(n)		
Space complexity : O(1)

where n is the size of the array
*/
